homework/index.js

```diff
@@ -29,18 +33,218 @@
     });
     return elem;
   }
+  // create contributions layout as cards
+  const createContributionsLayout = contributionsData => {
+    const mainParent = document.getElementById('main');
+    const contributionCardsContainer = createAndAppend('div', mainParent, {
+      id: 'repo-contributor',
+    });
+    const cardsGroup = createAndAppend('div', contributionCardsContainer, {
+      class: 'card-group d-flex container-fluid',
+    });
+    contributionsData.forEach(contribution => {
+      const card = createAndAppend('a', cardsGroup, {
+        class: 'card border border-danger shadow-lg mb-3 mx-2',
+        href: contribution.html_url,
+        target: '_blank',
+      });
+      createAndAppend('img', card, {
+        class: 'card-img-top',
+        src: contribution.avatar_url,
+        alt: `${contribution.login}'s avatar`,
+      });
+      const cardBody = createAndAppend('div', card, {
+        class: 'card-body text-center',
+      });
 
-  function main(url) {
-    fetchJSON(url, (err, repositories) => {
-      const root = document.getElementById('root');
-      if (err) {
-        createAndAppend('div', root, { text: err.message, class: 'alert-error' });
+      createAndAppend('h5', cardBody, {
+        class: 'card-title btn-primary',
+        text: contribution.login,
+      });
+      const contributionInfo = createAndAppend('p', cardBody, {
+        class: 'card-text btn btn-danger d-flex justify-content-between rounded-0',
+        text: 'Contributions:',
+      });
+      createAndAppend('span', contributionInfo, {
+        class: 'card-text badge badge-light d-flex justify-content-between ',
+        text: contribution.contributions,
+      });
+    });
+  };
+  // create repository layout which is under the header
+  const createRepositoryLayout = repositoriesData => {
+    const mainParent = document.getElementById('main');
+    const leftColumn = createAndAppend('div', mainParent, {
+      class: 'd-flex flex-column repo-detail',
+      id: 'repo-detail',
+    });
+    //
+    const divWidgetRepoName = createAndAppend('div', leftColumn, {
```

Comment:
Your four 'widget' rows are almost identical. In the spirit of _Don't Repeat Yourself_ (**DRY**) you could extract the common code into a function and call it four times, supplying the relevant data as parameters. For instance, you could create a function named `createRepositoryWidget` with the following declaration:

```js
function createRepositoryWidget(container, options = {}) {
  ...
}
```

You could then call that function like this, taking into consideration that the first widget need some extra work:

```js
const createRepositoryLayout = repositoriesData => {
  ...
  const firstWidget = createRepositoryWidget(leftColumn, {
    faClasses: 'fab fa-github',
    title: 'Repository Name:',
    valueTag: 'a',
    value: repositoriesData.name,
  });
  const anchorTag = firstWidget.lastChild;
  anchorTag.setAttribute('href', repositoriesData.html_url);
  anchorTag.setAttribute('target', '_blank');

  createRepositoryWidget(leftColumn, {
    faClasses: 'fas fa-pen-alt',
    title: 'Description',
    value: repositoriesData.description,
  });

  ...
}
```

I'm sure you will have no difficulty devising what the implementation should be of the `createRepositoryWidget` function should be.

Reply:
`const anchorTag = firstWidget.lastChild;` 
my solution is a little bit different because the funct i wrote does not return value thats why i use `const anchorTag = leftColumn.firstChild.lastChild;`

Also for the anchor tag creation i have used
 `  createAndAppend(options.valueTag || 'h4', widgetContainer, {
      text: options.value,
    ...
    });`

Reply:
Your `createRepositoryWidget` function could simply return the top-level container element of the widget as its return value:

```js
function createRepositoryWidget(container, options = {}) {
  ...
  return widgetContainer;
}


function createRepositoryLayout(repositoriesData) {
  ...
  const widgetContainer = createRepositoryWidget(leftColumn, {
    faClasses: 'fab fa-github',
    title: 'Repository Name:',
    value: repositoriesData.name,
    valueTag: 'a',
  });

  const anchorTag = widgetContainer.lastChild;
  anchorTag.setAttribute('href', repositoriesData.html_url);
  anchorTag.setAttribute('target', '_blank');
  ...
}
```

In other cases where you call `createRepositoryWidget()` and have no use for its return value, just ignore it (the same as is done with `createAndAppend()`).